src/Features/CSharp/Portable/CodeFixes/AddParameterXmlDocumentation/AddParameterXmlDocumentationCodeFixProvider.cs

```diff
@@ -0,0 +1,316 @@
+                            {
+                                _documentedParameterAfter = node;
+                                _documentedParameterAfterIndex = index;
+                            }
+                        }
+                    }
+                    else if (_documentedParameterAfter == null)
+                    {
+                        // in case all the param or typeparam tags that exist are not a parameter
+                        _documentedParameterAfter = node;
+                    }
+                }
+                base.VisitXmlNameAttribute(node);
+            }
+        }
+
+        private class AddParameterXmlDocumentationCodeAction : CodeAction
+        {
+            private string _title;
+            private Document _document;
+            private string _parameterName;
+            private XmlNameAttributeSyntax _adjacentAttribute;
+            private bool _adjacentAttributeIsBefore;
+            private bool _isTypeParameter;
+
+            public override string Title => _title;
+
+            public AddParameterXmlDocumentationCodeAction(Document document, string parameterName, XmlNameAttributeSyntax adjacentAttribute,
+                bool adjacentAttributeIsBefore, bool isTypeParameter)
+            {
+                if (isTypeParameter)
+                {
+                    _title = string.Format(CSharpFeaturesResources.AddTypeParameterXmlDocumentation, parameterName);
+                }
+                else
+                {
+                    _title = string.Format(CSharpFeaturesResources.AddParameterXmlDocumentation, parameterName);
+                }
+                _document = document;
+                _parameterName = parameterName;
+                _adjacentAttribute = adjacentAttribute;
+                _adjacentAttributeIsBefore = adjacentAttributeIsBefore;
+                _isTypeParameter = isTypeParameter;
+            }
+
+            protected override async Task<Document> GetChangedDocumentAsync(CancellationToken cancellationToken)
+            {
+                var root = await _document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+                // TODO Handle (or disallow) /** */ comments
+                var newLineAndSpace = GetXmlCommentNewLineAndSpace("///");
+                var elementName = _isTypeParameter ? "typeparam" : "param";
+                var newDocumentation = GetNewDocumentationParamElement(_parameterName, elementName, _adjacentAttribute);
+
+                if (_adjacentAttributeIsBefore)
+                {
+                    root = root.InsertNodesAfter(_adjacentAttribute.Parent.Parent,
+                        new SyntaxNode[] { newLineAndSpace, newDocumentation });
+                }
+                else
+                {
+                    root = root.InsertNodesBefore(_adjacentAttribute.Parent.Parent,
+                        new SyntaxNode[] { newDocumentation, newLineAndSpace });
+                }
```

Comment:
That's certainly cleaner, but I also need to call a different `InsertNodes...` method in the two branches.

Looking at this code again I realized that I introduced a bug here. If `_adjacentAttribute` is an `XmlEmptyElementSyntax` I only want `.Parent` once and not twice. At some point I was passing around the Xml element and the attribute separately and then I cleaned the element away...